// ufp-core/lib/modules/intl/IntlReducer.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _defineProperty2 = require('babel-runtime/helpers/defineProperty');

var _defineProperty3 = _interopRequireDefault(_defineProperty2);

var _ReduxUtils = require('../../utils/ReduxUtils');

var _ReduxUtils2 = _interopRequireDefault(_ReduxUtils);

var _IntlConstants = require('./IntlConstants');

var _IntlConstants2 = _interopRequireDefault(_IntlConstants);

var _IntlActionHandlers = require('./IntlActionHandlers');

var _IntlActionHandlers2 = _interopRequireDefault(_IntlActionHandlers);

var _IntlConfig = require('./IntlConfig');

var _IntlConfig2 = _interopRequireDefault(_IntlConfig);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

// import IntlActionCreators from './IntlActionCreators'
var initialState = {
    currentLanguage: _IntlConstants2.default.DEFAULT_LANGUAGE,
    randomKey: Math.random(),
    nextLanguage: null,
    locales: _IntlConfig2.default.getLocales(),
    allMessages: (0, _defineProperty3.default)({}, _IntlConstants2.default.DEFAULT_LANGUAGE, {}),
    languages: _IntlConfig2.default.getLanguages()
};

exports.default = _ReduxUtils2.default.createReducer(initialState, _IntlActionHandlers2.default);
