// jscpd/lib/map.js
// Generated by CoffeeScript 2.0.2
(function() {
  var Map;

  Map = class Map {
    constructor() {
      this.clones = [];
      this.clonesByFile = {};
      this.numberOfDuplication = 0;
      this.numberOfLines = 0;
      this.numberOfFiles = 0;
    }

    addClone(clone) {
      this.clones.push(clone);
      this.numberOfDuplication = this.numberOfDuplication + clone.linesCount;
      if (clone.firstFile in this.clonesByFile) {
        this.clonesByFile[clone.firstFile].push(clone.firstFile);
      } else {
        this.clonesByFile[clone.firstFile] = [clone.firstFile];
        this.numberOfFiles++;
      }
      if (clone.secondFile in this.clonesByFile) {
        return this.clonesByFile[clone.secondFile].push(clone);
      } else {
        this.clonesByFile[clone.secondFile] = [clone];
        return this.numberOfFiles++;
      }
    }

    getPercentage() {
      var result;
      result = 0;
      if (this.numberOfLines > 0) {
        result = this.numberOfDuplication / this.numberOfLines * 100;
      }
      return result.toFixed(2);
    }

  };

  exports.Map = Map;

}).call(this);
