// jscpd/lib/report.js
// Generated by CoffeeScript 2.0.2
(function() {
  var Report, fs, logger, path;

  fs = require('fs');

  logger = require('winston');

  path = require('path');

  Report = class Report {
    constructor(options) {
      var cwd, ext, isAbsolute, reporter;
      this.options = options;
      reporter = this.options.reporter;
      if (this.options.output) {
        ext = this.options.output.split('.').pop();
      }
      if (ext === 'xml' && reporter === 'json' || ext === 'json' && reporter === 'xml') {
        logger.warn(`output file extention '${this.options.output}' does not match reporter '${reporter}'`);
      }
      switch (reporter) {
        case 'xml':
          reporter = './reporters/xml-pmd';
          break;
        case 'json':
          reporter = './reporters/json';
          break;
        default:
          cwd = process.cwd();
          reporter = path.normalize(reporter);
          isAbsolute = reporter.indexOf(cwd) === 0;
          if (!isAbsolute) {
            reporter = path.resolve(cwd, reporter);
          }
      }
      this.reporter = require(reporter);
      this.stdReporter = require('./reporters/_std-log');
    }

    generate(map) {
      var dump, log, raw;
      this.map = map;
      [raw, dump, log] = this.reporter(this.options);
      if (!log) {
        log = this.stdReporter();
      }
      logger.info(log);
      if (this.options.output) {
        fs.writeFileSync(this.options.output, dump || raw);
      } else {
        logger.warn('output file is not provided');
      }
      return raw || dump;
    }

  };

  exports.Report = Report;

}).call(this);
